keps/sig-node/2823-pod-addmission-plugin/README.md

```diff
@@ -0,0 +1,445 @@
+    PluginTypeGRPC   PluginType = "grpc"   // Local port on the host.
+    PluginTypeSocket PluginType = "socket" // fd to connect to.
+)
+
+// AdmissionPluginManager is the podAdmitHandler shim for external plugins.
+type AdmissionPluginManager struct {
+    confDir      string
+    binDir       string
+    pluginConfig *PluginConfig
+}
+
+// PluginConfig represents the plugin configuration file
+type PluginConfig struct {
+    Name         string              `json:"name,omitempty"`
+    Version      string              `json:"version,omitempty"`
+    Plugins      []*AdmissionPlugin  `json:"plugins"`
+}
+
+// AdmissionPlugin represents individual plugins specified in the configuration.
+type AdmissionPlugin struct {
+    Name     string     `json:"name"`
+    Type     PluginType `json:"type"`
+}
+
+```
+
+#### Feature gate
+
+This functionality adds a new feature gate named “PodAdmissionPlugin” which
+decides whether to invoke admission plugin or not.
+
+#### Kubelet to pod admission plugin communication
+
+Kubelet will encode the pod spec and invoke each admission plugin. After
+decoding the pod spec, plugins can perform additional validations and return the
+encoded form of the struct mentioned below to kubelet to decide whether to admit
+the pod or not. For shell plugin type, request will be sent over stdin and
+response will be received over stdout and stderr.
+
+```
+AdmitResult { 
+   Admit bool 
+   Reason string 
+   Message string 
+}
+```
+
+#### Implementation detail
+
+As part of this implementation, new sub package will be added to
+pkg/kubelet/lifecycle. In-tree admission handler shim will be included in this
+package, which will be responsible for discovering and invoking the pod
+admission plugins available on the host to decide on whether to admit the pod or
+not.
+
+If the plugin does not respond or if it's crashing, then Kubelet will fail-close
+and pod will not be accepted by the kubelet. Because if kubelet ignores the
+response and schedules the pod then intention of not executing pod with specific
+needs on these worker nodes will be violated. These plugins doesn't mutate the
+Pod object and can be invoked in parallel since there is no dependency between
+them.
+
+These validations will be applied to both static pods which are managed by
+Kubelet and pods scheduled by control plane.
+
+#### Handling of rejected pods
+
+The pod will enter a Terminated state and by default won’t be deleted or
+rescheduled but this behavior is consistent with in-tree admission handler
+rejections and it’s well documented (e.g.,
+in [topology-manager-docs](https://kubernetes.io/docs/tasks/administer-cluster/topology-manager/#policy-single-numa-node)).
+
+One benefit of associating this rejection with a specific Reason provided by the
+plugins (and in-tree handlers) is that the cluster admin can choose to run an
+operator (e.g., the descheduler) that watches over pod rejections and delete the
+pods based on the Reason field.
+
+### Test Plan
```

Comment:
All these sections (Test Plan, Graduation Criteria, PRR review questionnaire) need to be filled.